docs/design/CONFIG_API.md

```diff
@@ -0,0 +1,339 @@
+* If data plane static configuration is required in the future, Envoy Gateway will use a separate file for this purpose.
+
+The `v1alpha1` version and `config.gateway.envoyproxy.io` API group get generated:
+```go
+// gateway/api/config/v1alpha1/doc.go
+
+// Package v1alpha1 contains API Schema definitions for the config.gateway.envoyproxy.io API group.
+//
+// +groupName=config.gateway.envoyproxy.io
+package v1alpha1
+```
+
+The initial `EnvoyGateway` API being proposed:
+```go
+// gateway/api/config/v1alpha1/envoygateway.go
+
+package valpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/gateway/pkg/provider/file"
+	"github.com/gateway/pkg/provider/kubernetes"
+)
+
+// EnvoyGateway is the Schema for the envoygateways API
+type EnvoyGateway struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// EnvoyGatewaySpec defines the desired state of Envoy Gateway.
+	EnvoyGatewaySpec `json:",inline"`
+}
+
+// EnvoyGatewaySpec defines the desired state of Envoy Gateway configuration.
+type EnvoyGatewaySpec struct {
+	// Gateway defines Gateway-API specific configuration. If unset, default
+	// configuration parameters will apply.
+	//
+	// +optional
+	Gateway *Gateway `json:"gateway,omitempty"`
+
+	// Provider defines the desired provider configuration. If unspecified,
+	// the Kubernetes provider is used with default parameters.
+	//
+	// +optional
+	Provider *Provider `json:"provider,omitempty"`
+}
+
+// Gateway defines desired Gateway API configuration of Envoy Gateway.
+type Gateway struct {
+	// ControllerName defines the name of the Gateway API controller. If unspecified,
+	// defaults to "gateway.envoyproxy.io/gatewayclass-controller". See the following
+	// for additional details:
+	//
+	// https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.GatewayClass
+	//
+	// +optional
+	ControllerName string `json:"controllerName,omitempty"`
+}
+
+// Provider defines the desired configuration of a provider.
+// +union
+type Provider struct {
+	// Type is the type of provider to use. If unset, the Kubernetes provider is used.
+	//
+	// +unionDiscriminator
+	Type ProviderType `json:"type,omitempty"`
+	// Kubernetes defines the configuration of the Kubernetes provider. Kubernetes
+	// provides runtime configuration via the Kubernetes API.
+	//
+	// +optional
+	Kubernetes *kubernetes.Provider `json:"kubernetes,omitempty"`
+
+	// File defines the configuration of the File provider. File provides runtime
+	// configuration defined by one or more files.
+	//
+	// +optional
+	File *file.Provider `json:"file,omitempty"`
+}
+
+// ProviderType defines the types of providers supported by Envoy Gateway.
+type ProviderType string
+
+const (
+	// KubernetesProviderType defines the "Kubernetes" provider.
+	KubernetesProviderType ProviderType = "Kubernetes"
+
+	// FileProviderType defines the "File" provider.
+	FileProviderType ProviderType = "File"
```

Comment:
nit: Should the values be lowercase?
```
provider:
  type: kubernetes
  kubernetes: {}
```

Reply:
+1 -- I think that if the `kubernetes` attribute name is lowercase, the `type` value should match.

Reply:
+1 to lowercase, mainly because it's easier to remember that everything on the right hand side is lower-case than to remember which things are which case.